### PR TITLE
integrate Channels-based WebSockets into SignalR

### PR DESCRIPTION
**Pull request by anurse**
No automated tests yet, blocked on #26 (and functional tests which come after that). Also, could refactor for better (read: any) unit testing, but I want to get this merged. Logging an issue to improve WebSocket testing.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs [hunk 1]
```diff
@@ -3,143 +3,163 @@
...
+        private readonly WebSocketOpcode _opcode;
+        private readonly ILogger _logger;
 
-        public WebSockets(Connection connection, Format format)
+        public WebSockets(Connection connection, Format format, ILoggerFactory loggerFactory)
         {
-            _connection = connection;
-            _channel = (HttpChannel)connection.Channel;
-            _messageType = format == Format.Binary ? WebSocketMessageType.Binary : WebSocketMessageType.Text;
+            _channel = (HttpChannel) connection.Channel;
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: nit: space

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs [hunk 2]
```diff
@@ -3,143 +3,163 @@
...
+                    sending);
 
-                if (result.MessageType != WebSocketMessageType.Close)
+                // What happened?
+                if (trigger == receiving)
                 {
-                    outputBuffer.Advance(result.Count);
+                    // Shutting down because we received a close frame from the client.
+                    // Complete the input writer so that the application knows there won't be any more input.
+                    _logger.LogDebug("Client closed connection. Signalling end-of-input to application");
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: Are you able to log closing status?

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs [hunk 3]
```diff
@@ -3,143 +3,163 @@
...
+            _opcode = format == Format.Binary ? WebSocketOpcode.Binary : WebSocketOpcode.Text;
+
+            _logger = (ILogger) loggerFactory?.CreateLogger<WebSockets>() ?? NullLogger.Instance;
         }
 
         public async Task ProcessRequestAsync(HttpContext context)
         {
-            if (!context.WebSockets.IsWebSocketRequest)
+            var feature = context.Features.Get<IHttpWebSocketConnectionFeature>();
+            if (feature == null || !feature.IsWebSocketRequest)
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: It has to be `feature?.IsWebSocketRequest ?? false`, which just seems more verbose to me.

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs [hunk 4]
```diff
@@ -3,143 +3,163 @@
...
-                }
-            }
-
-            // REVIEW: Should this ever happen?
-            if (!IsClosedOrClosedSent(ws))
-            {
-                // Close the output
-                await ws.CloseOutputAsync(WebSocketCloseStatus.NormalClosure, "", CancellationToken.None);
-            }
+            // Flush
```
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: remove

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs (hunk 3 above)
Comment (moozzyk) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: don't bother then

**Comment by moozzyk** [comment]
Looks good to me

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs [hunk 5]
```diff
@@ -3,143 +3,163 @@
...
                 {
-                    break;
+                    // The application finished sending. We're not going to keep the connection open,
+                    // so close it and wait for the client to ack the close
+                    _channel.Input.CompleteWriter();
+                    _logger.LogDebug("Application finished sending. Sending close frame.");
+                    await ws.CloseAsync(WebSocketCloseStatus.NormalClosure);
+
+                    _logger.LogDebug("Waiting for the client to close the socket");
+                    // TODO: Timeout.
```
Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: https://github.com/aspnet/SignalR/issues/31

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs [hunk 6]
```diff
@@ -3,142 +3,162 @@
...
         {
-            var webSocketState = GetWebSocketState(webSocket);
-
-            return webSocketState == WebSocketState.Closed ||
-                   webSocketState == WebSocketState.CloseSent ||
-                   webSocketState == WebSocketState.Aborted;
+            if (_logger.IsEnabled(LogLevel.Debug))
+            {
+                _logger.LogDebug(
+                    $"{action} frame: Opcode={frame.Opcode}, Fin={frame.EndOfMessage}, Payload={frame.Payload.Length}b");
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: Is the `b` at the end on purpose?

Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: It's for `bytes`. Clarified.

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs [hunk 7]
```diff
@@ -2,143 +2,161 @@
...
-        public WebSockets(Connection connection, Format format)
+        public WebSockets(Connection connection, Format format, ILoggerFactory loggerFactory)
         {
-            _connection = connection;
-            _channel = (HttpChannel)connection.Channel;
-            _messageType = format == Format.Binary ? WebSocketMessageType.Binary : WebSocketMessageType.Text;
+            _channel = (HttpChannel) connection.Channel;
+            _opcode = format == Format.Binary ? WebSocketOpcode.Binary : WebSocketOpcode.Text;
+
+            _logger = (ILogger) loggerFactory?.CreateLogger<WebSockets>() ?? NullLogger.Instance;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: nit: space here too

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs [hunk 8]
```diff
@@ -2,143 +2,155 @@
...
+                _logger.LogInformation("Socket opened.");
 
-                ArraySegment<byte> segment;
-                if (!outputBuffer.Memory.TryGetArray(out segment))
-                {
-                    // REVIEW: Do we care about native buffers here?
-                    throw new InvalidOperationException("Managed buffers are required for Web Socket API");
-                }
+                // Begin sending and receiving. Receiving must be started first because ExecuteAsync enables SendAsync.
+                var receiving = ws.ExecuteAsync((frame, self) => ((WebSockets) self).HandleFrame(frame), this);
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: nit: less space :trollface:

File: src/Microsoft.AspNetCore.Sockets/WebSockets.cs [hunk 9]
```diff
@@ -2,143 +2,155 @@
...
-using System.Net.WebSockets;
-using System.Threading;
 using System.Threading.Tasks;
 using Channels;
 using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.WebSockets.Internal;
+using Microsoft.Extensions.Internal;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.Extensions.WebSockets.Internal;
```
Comment (BrennanConroy) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: Are both these WebSockets namespaces needed?

Comment (anurse) [inline src/Microsoft.AspNetCore.Sockets/WebSockets.cs]: Yes, Extensions has the IWebSocketConnection and AspNetCore has IHttpWebSocketConnectionFeature.

**Comment by BrennanConroy** [comment]
Looks good except for some of those commit messages

**Comment by anurse** [comment]
That's what squashing is for :). Is someone going to actually Approve the changes?

**Pull request merged**